src/client.cpp

```diff
@@ -1155,13 +1165,49 @@ void Client::sendInventoryAction(InventoryAction *a)
 	Send(&pkt);
 }
 
+bool Client::canSendChatMessage() const
+{
+	u32 now = time(NULL);
+	float time_passed = now - m_last_chat_message_sent;
+
+	float virt_chat_message_allowance = m_chat_message_allowance + time_passed * (CLIENT_CHAT_MESSAGE_LIMIT_PER_10S / 8.0f);
+
+	if (virt_chat_message_allowance < 1.0f)
+		return false;
+
+	return true;
+}
+
 void Client::sendChatMessage(const std::wstring &message)
 {
-	NetworkPacket pkt(TOSERVER_CHAT_MESSAGE, 2 + message.size() * sizeof(u16));
+	const s16 max_queue_size = g_settings->getS16("max_out_chat_queue_size");
+	if (canSendChatMessage()) {
+		u32 now = time(NULL);
+		float time_passed = now - m_last_chat_message_sent;
+		m_last_chat_message_sent = time(NULL);
 
-	pkt << message;
+		m_chat_message_allowance += time_passed * (CLIENT_CHAT_MESSAGE_LIMIT_PER_10S / 8.0f);
```

Comment:
I'm wondering, too. @nerzhul wrote this algorithm on server-side (https://github.com/minetest/minetest/blob/master/src/remoteplayer.cpp#L199)